### PR TITLE
Keyspace type specific events

### PR DESCRIPTION
**Pull request by inponomarev**
Adds the type-specific keyspace-notification event classes, completing the bulk of #406. Builds on #811, which delivered the configuration parameter, the two channel families and the generic (`g`) and expired (`x`) classes.

## What this adds

| Flag | Class | Events |
| --- | --- | --- |
| `$` | string | `set`, `append`, `setrange`, `incrby`, `incrbyfloat` |
| `l` | list | `lpush`, `rpush`, `lpop`, `rpop`, `linsert`, `lset`, `lrem`, `ltrim`, `sortstore` |
| `s` | set | `sadd`, `srem`, `spop`, `sinterstore`, `sunionstore`, `sdiffstore` |
| `h` | hash | `hset`, `hdel`, `hincrby`, `hincrbyfloat` |
| `z` | sorted set | `zadd`, `zincr`, `zrem`, `zremrangeby{score,rank,lex}`, `z{union,inter,diff}store`, `zrangestore`, `zpopmin`, `zpopmax` |
| `t` | stream | `xadd`, `xtrim`, `xdel` |
| `n` | new key | `new` |

Together with #811 that covers every class Jedis-Mock can support. The README gains a "Keyspace notifications" section documenting the flags, the events, and the remaining limitations.

## How the event names were established

Every name and ordering was derived by probing `redis:7.4-alpine` and then pinned with a comparison test, rather than inferred from the command name. That mattered — the following are all counter-intuitive and all verified:

* `DECR` and `DECRBY` report **`incrby`**, not `decrby`.
* `ZINCRBY` reports **`zincr`**; so does `ZADD … INCR`.
* `UNLINK` reports **`del`**, not `unlink`.
* `HSETNX` and `HMSET` both report **`hset`**.
* A multi-element push reports **one** event, not one per element; `LPUSHX`/`RPUSHX` report the same event as their unconditional forms.
* Emptying a container reports the **generic** `del` in addition to the type's own event — the `del` is not the container's class.
* `new` precedes the type's own event, and is **not** covered by the `A` alias.
* Orderings are not consistent between commands: `RPOPLPUSH` reports the destination's `lpush` *before* the source's `rpop`, whereas `SMOVE` reports the source's `srem` **and the `del` for the set it emptied** before the destination's `sadd`.
* `XCLAIM`/`XAUTOCLAIM` publish no event of their own in 7.4 — only `xgroup-createconsumer`, and only for a consumer that did not exist.

## Bugs found while self-reviewing this branch

Each was first reproduced as a comparison test that fails against Jedis-Mock and passes against real Redis:

1. **`ZADD key INCR score member` reported `zadd`** instead of `zincr`.
2. **`SMOVE key key member` destroyed and recreated the set.** Redis short-circuits when source and destination are the same key and simply reports membership; the mock ran `SREM` then `SADD`, so a single-member set was deleted and recreated (and published `srem`/`del`/`sadd`). This was a **data-correctness bug predating this branch**, not only a notification one.
3. **`SETRANGE` writing identical bytes published nothing** — Redis writes, and reports, whenever the value argument is non-empty.
4. **`ZRANGESTORE` from a missing source did not publish `del`** for a destination that existed; the early-return path bypassed the notification.
5. **A no-op `LTRIM` published nothing and did not dirty the key for `WATCH`.** Unlike `LREM`, Redis signals `LTRIM` as a modification unconditionally, even when the range covers the whole list.

Three further edge cases turned out to already behave correctly — an empty-value `SETRANGE` (silent), an empty-value `APPEND` (reports), and `RPOPLPUSH` rotating a list onto itself — and their tests are kept as regression guards.

## Reviewer note: `WATCH` semantics change

Six commands that mutate a value in place rather than replacing it — `LINSERT`, `LSET`, `LREM`, `LTRIM`, `XDEL`, `XTRIM` — never marked their key as modified, so a transaction watching that key was **not** aborted. Publishing notifications required fixing that, so this branch changes `WATCH` behaviour for those commands. It is a fix (it matches real Redis), but it is worth reviewing deliberately rather than as a side effect. `WatchInPlaceMutationTest` covers both directions: real mutations abort a watching transaction, a no-op `LREM` does not, and a no-op `LTRIM` does.

## Testing

* **59 comparison tests** in `comparisontests/notifications`, each executed against both Jedis-Mock and real Redis, asserting exact event *sequences* including the keyspace/keyevent split, class masking, cross-database delivery and empty-container `del`.
* **12 upstream keyspace tests in `unit/pubsub.tcl` now run and pass** (previously all disabled): the `KA`/`EA`/`KEA` trio, event masking, general, list, set, zset, hash, triggered expiry, CONFIG flag round-tripping and new-key.
* Full Java suite and the whole native-test matrix green.

## Still out of scope

Four upstream tests remain disabled, each with the reason in a comment beside it:

* **`expired events (background expire)`** — Jedis-Mock expires keys lazily, so a key nobody touches produces its `expired` event later than a real server would. An active-expiry cycle is the natural follow-up.
* **`expired event (Expiration time is already expired)`** — an expiration set in the past reports `del` on Redis and `expired` on Valkey; this follows Redis, matching the container the comparison tests run against.
* **`evicted events`** — there is no `maxmemory` model, so nothing is ever evicted.
* **`stream events test`** — it drives `XGROUP`/`XREADGROUP`/`XCLAIM`/`XAUTOCLAIM`, which Jedis-Mock does not implement at all, so their `xgroup-*` events cannot exist.

Also unimplemented: the `m` (key-miss) and `d` (module) classes, and `expired` from a `TTL`/`PTTL` lookup on an already-expired key (the event is published from the value-access path).

## Possible follow-up

The "type event, then generic `del` if the container is now gone" logic is repeated at eight call sites using three different existence checks (`list.isEmpty()`, `set.isEmpty()`, `!base().exists(key)`), which disagree about which of them deletes the key. Three of the five bugs above lived in that inconsistency. Consolidating it behind a single helper on `RedisBase` looks worthwhile, with one trap to watch: `exists()` runs `verifyKey`, which can itself publish `expired`.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

File: src/main/java/com/github/fppt/jedismock/operations/sortedsets/ZPop.java [hunk 1]
```diff
@@ -50,8 +51,11 @@ protected List<Slice> pop() {
                     .collect(Collectors.toList()));
             mapDBObj.remove(entry.getValue());
         }
+        //One event per command, then the generic del if the set is now gone
+        base().notifyKeyspaceEvent(isRev ? KeyspaceEvent.ZPOPMAX : KeyspaceEvent.ZPOPMIN, key);
         if (mapDBObj.isEmpty()) {
             base().deleteValue(key);
+            base().notifyKeyspaceEvent(KeyspaceEvent.DEL, key);
         }
```
[the remark this answers is not in the document]
Comment (inponomarev) [reply src/main/java/com/github/fppt/jedismock/operations/sortedsets/ZPop.java]: fixed

File: src/main/java/com/github/fppt/jedismock/storage/ExpiringKeyValueStorage.java [hunk 2]
```diff
@@ -131,6 +150,7 @@ public void put(Slice key1, Slice key2, Slice value, Long ttl) {
         RMHash mapByKey;
 
         if (!values.containsKey(key1)) {
+            eventPublisher.accept(KeyspaceEvent.NEW, key1);
             mapByKey = new RMHash(getClockSupplier());
             values.put(key1, mapByKey);
         } else {
```
[the remark this answers is not in the document]
Comment (inponomarev) [reply src/main/java/com/github/fppt/jedismock/storage/ExpiringKeyValueStorage.java:172]: not reproducible. `writingAFieldOfAnExpiredHashPublishesExpiredThenNew` passes on the mock unmodified. Every caller of that overload (HSet, HSetNX, HIncrBy) reads the old field first via base().getSlice(key1, key2), which runs verifyKey and expires the key before the write ever happens. So the bug was already masked. I applied expireIfOutdated there anyway as defence in depth, and the test stays as a regression guard in case a future caller writes without reading first.
